pkg/scheduler/preemption/classical/candidate_generator.go

```diff
@@ -0,0 +1,165 @@
+/*
+Copyright The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package classical
+
+import (
+	"sort"
+	"time"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/utils/clock"
+
+	kueue "sigs.k8s.io/kueue/apis/kueue/v1beta1"
+	"sigs.k8s.io/kueue/pkg/cache"
+	"sigs.k8s.io/kueue/pkg/resources"
+	"sigs.k8s.io/kueue/pkg/workload"
+)
+
+type candidateIterator struct {
+	candidates                         []*candidateElem
+	RunIndex                           int
+	frsNeedPreemption                  sets.Set[resources.FlavorResource]
+	snapshot                           *cache.Snapshot
+	AnyCandidateFromOtherQueues        bool
+	AnyCandidateForHierarchicalReclaim bool
+	hierarchicalReclaimCtx             *HierarchicalPreemptionCtx
+}
+
+type candidateElem struct {
+	wl *workload.Info
+	// lca of this queue and cq (queue to which the new workload is submitted)
+	lca *cache.CohortSnapshot
+	// candidates above priority threshold cannot be preempted if at the same time
+	// cq would borrow from other queues/cohorts
+	onlyWithoutBorrowing bool
+	reclaim              bool
```

Comment:
can this be simplified to an enumeration? Conceptually, I think we have 4 cases:

1) Non-preemptable
2) within CQ
3) reclaim
4) borrowWithinCohort

We only need to represent the latter 3 here.

Reply:
this would also allow some other simplification - we can filter all candidates from list 4 when borrow=false

`getPreemptionReason` would be just a simple mapping of preemption type to string